### DISCUSSION TITLE
Faster conversion from a bunch of vtk files to images using pyvista

### DISCUSSION THREAD

**Discussion by mellow753** [pyvista/pyvista#2324] [Q&A] [answered]
Hi, I have thousands of vtk files, and I use pyvista convert them to png images. 
The main body in python code is shown below:
```
import pyvista as pv
import os

path = 'C:/Users/su/Desktop/vtk//Ur15/'
vtkFiles= os.listdir(path_vtk)

path_img = path + 'image/'
os.makedirs(path_img)

for file in vtkFiles:

    vtkData = pv.read(path_vtk + file)
    mesh = pv.PolyData(vtkData)


    plotter = pv.Plotter(off_screen=True)
    plotter.add_mesh(mesh, scalars=field, cmap="jet")
    plotter.view_xy()
    plotter.screenshot((path_img + 'U.png'), transparent_background=True, window_size=[400,250])
```
But the for loop is has very low efficiency to convert the vtk file to image, around 1 second for 1 pic. So it will take more than 1 hour to complete the conversion for all vtk files.

**So I am wondering is there any way to improve the convert efficiency using python scripting?**

I appreciate for any suggrestion!

**Comment by RichardScottOZ** [comment, marked as answer]
Run it in parallel - map your list to multiprocessing.

**Reply by mellow753** [reply]
Thanks for your suggestion! Could you please give me a sample code for parallel python calculation? I am just starting to learn python, so still not very familiar with it.

**Reply by mellow753** [reply]
Hi @RichardScottOZ , I have tried the parallel calculation using `multiprocessing.Pool()` and `pool.map()` as you suggested. Thank you so much! I define `vtkToImage(filelist, inpath, outpath)` function to convert vtk file to image, however, the input of function is not only file list but also includes other argus, like `inpath` (path of vtk files) and `outpath` (path of images), and these two argus are the same for all sub-processors. Do you know how to deal with the multiple run function when parallel calculation? Thanks!

**Comment by RichardScottOZ** [comment]
You need a list where each list item has the multiple things you can use.

**Reply by RichardScottOZ** [reply]
runlist = [[filelist1, inputpath1,outputpath1],.....]
Then submit vtkToImage, idx[0], idx[1],..: idx for idx in runlist 

Type pattern for the particular syntax for your map

**Reply by mellow753** [reply]
Hi @RichardScottOZ , I test 40 vtk files mapping in 4 processors with the starmap() to deal with the multiple input of func, but I receive the error. I guess there is something wrong in my param_dict. 4 processos for 4 file chunks (10 files for each) have the same inpath and outpath.

This is my code
```
if __name__ == "__main__":
   
    num_cpus = 4
   
   # obtain filelist
   vtkFiles = getFile(path_vtk)

   # split filelist into chunks for multi-processoring
   filechunk = list()
   chunk_size = int(len(vtkFiles)/num_cpus)
   for i in range(0, len(vtkFiles), chunk_size):
       filechunk.append(vtkFiles[i:i+chunk_size])
   
   # build pool
   pool = mp.Pool(num_cpus)
   param_dict = [[filechunk[0], path_vtk, path_img],
                 [filechunk[1], path_vtk, path_img],
                 [filechunk[2], path_vtk, path_img],
                 [filechunk[3], path_vtk, path_img]]

   output = pool.starmap(vtkToImage, param_dict)
```

**Comment by RichardScottOZ** [comment]
What is the error?

**Reply by mellow753** [reply]
I call for split() in getFile() function, and the error is: `AttributeError: 'list' object has no attribute 'split'`. I guess 

The definition of getFile() is 
```
def getFile(path):
    filelist = []
    files = os.listdir(path)
    filelist = sorted(files, key=lambda x: float((x.split('_')[2])[:-4]))
    return filelist
```

**Comment by RichardScottOZ** [comment]
Yeah, not a multiprocessing error then, what does your list of filenames look like?  You are trying to get the index 2 of your split on underscores and then part of the string after that?

If you are just starting, doing many thing at once sorting, lambda function list comprehensions is likely to be confusing and will be plenty of mistakes in the process, so break it down until you get each part right - simple sort how you want, try splitting one file how you want, then select the piece and the substring and put it back together

**Reply by RichardScottOZ** [reply]
That will show you were the parenthesis in the wrong place is, etc.

**Reply by mellow753** [reply]
Hi @RichardScottOZ Many thanks for your suggestion! I have solved this problem by using the following code:
```
pool = mp.Pool(num_cpus)
pool.starmap(vtkToImage, [(vtkFiles[i], path_vtk, path_img) for i in range(len(vtkFiles))])   
```

Thanks for your help!

**Comment by RichardScottOZ** [comment]
Great - how much of a speedup did you get?

**Reply by mellow753** [reply]
1 second for 1 image for serial
1 second for 4 images for parallel (4 cpus)
The parallel running significantly speeds up the scripting!

**Comment by RichardScottOZ** [comment]
Linear is good, yes.
